src/main/java/com/nike/cerberus/config/MetricsConfigurationHelper.java

```diff
@@ -0,0 +1,210 @@
+package com.nike.cerberus.config;
+
+import com.amazonaws.util.EC2MetadataUtils;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.SlidingTimeWindowReservoir;
+import com.codahale.metrics.Timer;
+import com.google.inject.Inject;
+import com.nike.internal.util.Pair;
+import com.nike.riposte.metrics.codahale.CodahaleMetricsCollector;
+import com.nike.riposte.metrics.codahale.CodahaleMetricsListener;
+import com.nike.riposte.metrics.codahale.EndpointMetricsHandler;
+import com.nike.riposte.metrics.codahale.contrib.SignalFxReporterFactory;
+import com.nike.riposte.metrics.codahale.impl.SignalFxEndpointMetricsHandler;
+import com.nike.riposte.metrics.codahale.impl.SignalFxEndpointMetricsHandler.MetricDimensionConfigurator;
+import com.nike.riposte.server.http.HttpProcessingState;
+import com.nike.riposte.server.http.RequestInfo;
+import com.nike.riposte.server.http.ResponseInfo;
+import com.signalfx.codahale.reporter.SignalFxReporter;
+import com.signalfx.codahale.reporter.SignalFxReporter.Builder;
+import com.signalfx.codahale.reporter.SignalFxReporter.MetricDetails;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import static com.nike.riposte.metrics.codahale.impl.SignalFxEndpointMetricsHandler.DEFAULT_REQUEST_LATENCY_TIMER_DIMENSION_CONFIGURATOR;
+
+/**
+ * A set of methods used to help instantiate the necessary metrics objects in the Guice module
+ */
+public class MetricsConfigurationHelper {
+
+    private static final String EC2_HOSTNAME_PREFIX = "ip-";
+
+    private final String signalFxApiKey;
+
+    private final String signalFxAppEnvDim;
+
+    /**
+     * This 'holder' class allows optional injection of SignalFx-specific properties that are only necessary when
+     * SignalFx metrics reporting is enabled.
+     *
+     * The 'optional=true' parameter to Guice @Inject cannot be used in combination with the @Provides annotation
+     * or with constructor injection.
+     *
+     * https://github.com/google/guice/wiki/FrequentlyAskedQuestions
+     */
+    static class SignalFxOptionalPropertyHolder {
+        @Inject(optional=true)
```

Comment:
I'm pretty sure you can inject optional stuff using `@Nullable`, even into constructors or `@Provides` methods. Try the standard `javax.inject.Inject` and `javax.inject.Named` annotations rather than the guice-specific versions, and also mark it with `@Nullable`. If that doesn't work let me know and I'll mess around with it - I'm confident I've done this before.

Reply:
Also note that there are sometimes several `@Nullable` annotations available, and that whichever one you use must have `@Retention(RetentionPolicy.RUNTIME)` in order for Guice to be able to see it. When in doubt, use `javax.annotation.Nullable`.

Reply:
Further info here: https://github.com/google/guice/wiki/UseNullable

Reply:
Hmm, more reading makes it look like it still might not work. Worth a shot.

Reply:
@nicmunroe I thought so too, hoping that Guice would intelligently inject `null` if a property key does not exist in the `Named` map, but my testing showed that Guice still errors with `No implementation bound` even when @Nullable is applied.

Unfortunately `@Nullable` does not substitute for having some sort of provider, binding, or `Named` value declaration that tells Guice how to inject that variable.